test/rules/invariant_boolean.dart

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `dart test/util/solo_test.dart invariant_boolean`
+
+int bar = 1;
+
+int baz = 2;
+
+int foo = 0;
+bool setting = null;
+
+void bad() {
+  if ((foo == bar && someComputation()) || (foo != bar && otherComputation())) {} // OK
+  if (foo > bar || foo == bar) {} // OK
+  if (foo > bar && someComputation() || foo == bar) {} // OK
+  if (foo > bar || foo == bar && someComputation()) {} // OK
+  if (foo < bar - baz || foo > bar) {} // OK
+  if (setting == true) {} // OK
+  if (setting != true) {} // OK
+  if (foo == bar || foo != bar) {} // LINT
+  if (foo == bar || bar != foo) {} // LINT
+  if (foo == bar && foo != bar) {} // LINT
+  if (foo == bar && someComputation() && foo != bar) {} // LINT
```

Comment:
Same as above, I would argue that side effects are out of scope. I have seen this kind of rule to warn against residual debugging code, e.g. when forcing a branch of an if statement to reproduce/fix a bug and after fixing it, leaving the invariant condition.